Changes to personal goals

CSCMe opened the pull request:
🆑 CSCMe
tweak: Clicking "Show Goals" tells you that they are disabled in preferences if they are.
tweak: Jobs allow 0 to 5 personal goals by default, instead of 1 to 3.
bugfix: Removing/re-rolling goals acts on the correct goals if used more than once.
bugfix: You cannot roll the same goal twice in the same set, or no goal if re-rolling one.
/🆑


Other changes:
You can now see the goal that was added to you when a new one is generated.
"Add random new goal" message now disappears when you have the maximum amount of goals your job allows instead of just 5
"Add random new goal" message also disappears if your preferences don't allow you to have personal goals
Ghosts can no longer remove their personal goals

Admin stuff:
Adding a goal to someone else now shows you the goal you generated
More admin logging for when someone adds/removes/re-rolls goals for someone else
Admins are exempt from any restrictions about preferences and min/max goal amounts when meddling with someone's goals

MuckerMayhem commented on code/datums/mind/mind.dm:
```diff
@@ -162,36 +159,38 @@
-				to_chat(usr, SPAN_NOTICE("<b>You have removed a goal from \the [current]:</b> '[goal.summarize(FALSE, FALSE)]'."))
-				to_chat(current, SPAN_NOTICE("<b>A goal has been removed:</b> '[goal.summarize(FALSE, FALSE)]'."))
-			qdel(goal)
+		if(goal && (goal in goals) && can_modify)
```
```suggestion
		if(can_modify && goal && (goal in goals))
```
Try to keep the cheaper operations first to save resources.

MuckerMayhem commented on code/datums/mind/mind.dm:
```diff
@@ -162,36 +159,38 @@
-		if(caller && caller == current) can_modify = TRUE
+		if(caller && caller == current && !isghost(usr)) can_modify = TRUE
 
 		if(goal && (goal in goals) && can_modify)
```
Ditto.

MuckerMayhem commented on code/modules/goals/goal_mind.dm:
```diff
@@ -13,39 +13,67 @@
+/datum/mind/proc/delete_goal(var/datum/job/job, var/datum/goal/goal, var/override_min_goals = FALSE)
+	var/min_goals = job ? job.min_goals : 1
+
+	if(LAZYLEN(goals) == min_goals && !override_min_goals)
```
```suggestion
	if(!override_min_goals && LAZYLEN(goals) == min_goals)
```

MuckerMayhem commented on code/modules/goals/goal_mind.dm:
```diff
@@ -13,39 +13,67 @@
+	var/min_goals = job ? job.min_goals : 1
+
+	if(LAZYLEN(goals) == min_goals && !override_min_goals)
+		to_chat(src.current, SPAN_WARNING("Your job needs you to have at least a few distractions."))
```
Just a nitpick, but you may want to change this to say `at least one distraction` since the fallback count for `min_goals` is `1`.

Spookerton commented on code/modules/goals/goal_mind.dm:
```diff
@@ -13,39 +13,67 @@
 		new goal(src)
 	return TRUE
+
+/datum/mind/proc/delete_goal(var/datum/job/job, var/datum/goal/goal, var/override_min_goals = FALSE)
```
```suggestion
/datum/mind/proc/delete_goal(datum/job/job, datum/goal/goal, override_min_goals)
```

Spookerton reviewed (changes requested):
no var/ in new/changed signatures, null evaluates false

Spookerton commented on code/modules/goals/goal_mind.dm:
```diff
@@ -13,39 +13,67 @@
 
 // Create and display personal goals for this round.
-/datum/mind/proc/generate_goals(var/datum/job/job, var/adding_goals = FALSE, var/add_amount, var/is_spawning = FALSE)
+/datum/mind/proc/generate_goals(var/datum/job/job, var/adding_goals = FALSE, var/add_amount, var/is_spawning = FALSE, var/bypass_goal_checks = FALSE)
```
```suggestion
/datum/mind/proc/generate_goals(datum/job/job, adding_goals, add_amount, is_spawning, bypass_goal_checks)
```

CSCMe committed:
Changes to personal goals

You can now no longer have the same goal twice.

You can now see the goal that was added to you when you generate a new one.

Your job now specifies how many personal goals you can have (default: 0-5)

The game now tells you when your preferences are set so you can't generate goals when you click on "Show Goals"

More admin logging for adding/removing/rerolling goals

Fixed rerolled goals sometimes disappearing without generating a new one  

Fixed removing/rerolling goals sometimes removing/rerolling the wrong goals

Spookerton commented:
unrelated but @Spookerton  todo: gate add random goal in topic, since you can add as many as you like so long as you use an existing link. <https://i.imgur.com/N2jUyai.png>
